src/unleash.ts

```diff
@@ -34,32 +44,67 @@ export class Unleash extends EventEmitter {
             throw new Error('Unleash server URL missing');
         }
 
-        this.repository = new Repository(backupPath, url, refreshInterval);
+        if (!appName) {
+            throw new Error('Unleash client appName missing');
+        }
 
-        this.repository.on('error', (err) => {
-            this.emit('error', err);
-        });
+        if (!instanceId) {
+            instanceId = `generated-${Math.round(Math.random() * 1000000)}-${process.pid}`;
+        }
+
+        const requestId: string = `${appName}:${instanceId}`;
+        this.repository = new Repository(backupPath, url, requestId, refreshInterval);
 
         strategies = [new Strategy('default', true)].concat(strategies);
 
         this.repository.on('ready', () => {
             this.client = new Client(this.repository, strategies, errorHandler);
             this.emit('ready');
         });
+
+        this.repository.on('error', (err) => {
+            err.message = `Unleash Repository error: ${err.message}`;
+            this.emit('error', err);
+        });
+
+        this.repository.on('warn', (msg) => {
+            this.emit('warn', msg);
+        });
+
+        this.metrics = new Metrics({
+            disableMetrics,
+            appName,
+            instanceId,
+            strategies: strategies.map((strategy: Strategy) => strategy.name),
+            metricsInterval,
+            url
+        });
+
+        this.metrics.on('error', (err) => {
```

Comment:
still missing an `this.metrics.on(''count')` so we can also chuck this in the consumers of the client. Seems like only I want that though